### DISCUSSION TITLE
Packaged WinUI app crashes on start-up when deployed to test PC

### DISCUSSION THREAD

**Discussion by Slion** [microsoft/WindowsAppSDK#2407] [Dev team discussions]
That app works great on both my development machines.
However it failed Windows Store submission because it crashed on start-up during certification.

Sure enough it crashes on start-up when installed on a third PC with something similar to that:
https://stackoverflow.com/questions/45200620/faulting-module-name-kernelbase-dll
Event logs reports: `Faulting module name: KERNELBASE.dll`

I've added logs and registered to that `Application.UnhandledException` event however I'm not getting any logs meaning it looks like it crashes before the constructor of my Application class implementation.
Is there an actual entry point I could use outside the Application class?

Updated to 1.1.0-preview1, same issue.

I've also determined that the application starts properly since some of my code is running. I can see the blank window with size and title bar text and icon set as expected but then it crashes badly later on and NLog does not even get a chance to write its logs somehow.
Even forcing flush with NLog target set to a file did not work. For some reason the log file is not created on the test PC even though it works fine on my development machine. Fortunately with NLog set to send logs over UDP using NLogViewer target I can get the logs using [Sentinel](https://github.com/yarseyah/sentinel).

Looks more and more like my issue has little to do with Windows App SDK after all. Still how comes I'm not getting unhandled exception event? Though the app works on that third PC when running from Visual Studio but still does not run when installed from an installer.

Related issues:
- https://github.com/microsoft/WindowsAppSDK/issues/2346 - UWP
- https://github.com/microsoft/WindowsAppSDK/issues/2117

**Comment by Slion** [comment]
My issue was caused by trying to write files to `AppDomain.CurrentDomain.BaseDirectory`.  Switching to `Windows.Storage.ApplicationData.Current.LocalCacheFolder.Path` fixed it.

What's odd is that on some Windows installation it works and on others it does not…
That also explains why NLog could not create its file next to the executable.
When using NLog with packaged app you should have a target that looks like:
```XML
<!-- That logs to something like: C:\Users\micro\AppData\Local\Packages\59798Slions.TaskbarPro_md9twrekhzh1t\LocalCache\Local\logs.txt --> 
<target name="File" xsi:type="File" fileName="${specialfolder:folder=LocalApplicationData:cached=true}/logs.txt" />
```

Just make sure your packaged app does not try to write to files relative to your executable location.

As to why the unhandled exception event is not called and why `KERNELBASE.dll` is flagged as causing the issue, it could be because the exception was triggered from the callback from [`EnumWindows`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumwindows).

**Reply by Slion** [reply]
> What's odd is that on some Windows installation it works and on others it does not…

I wonder if [Test Mode](https://slions.net/threads/toggle-windows-test-mode-and-kernel-debugging.88/) may have something to do with that. Both machines where writing to `AppDomain.CurrentDomain.BaseDirectory` worked had Test Mode enabled.

Moreover when running your app from Visual Studio writing to `AppDomain.CurrentDomain.BaseDirectory` also works which does not help spotting those issues during development.

**Reply by DrusTheAxe** [reply]
BCDEDIT's testsigning and debug won't affect the pkgdir's un-writeability.

>when running your app from Visual Studio

That'll work, because a package's installed location varies, depending...

VS F5 registers a package with pkgdir=`$(OutDir)` which is of course writeable, whereas PackageManager.AddPackage*() adds the package under a package volume which are locked down without write access.

The longer explanation...

Ultimately all MSIX deployment operations start with a call to [PackageManager](https://docs.microsoft.com/uwp/api/windows.management.deployment.packagemanager?view=winrt-22000)

PackageManager.AddPackage*() creates a pkgdir as the package's full name (`LolzCatzVideoz_1.2.3.4_x86__1234567890abc` etc) under the current default package volume (default is `C:\Program Files\WindowsApps` but can be changed). Deployment owns this location.

PackageManager.AddPackage*(...targetVolume=[packagevolume](https://docs.microsoft.com/uwp/api/windows.management.deployment.packagevolume?view=winrt-22000)...) creates a pkgdir as the package's full name under the specified package volume. Deployment owns all package volumes so Windows is responsible for lifetime, access, etc.

A new Windows image starts with the default package volume but you can add more (via PackageManager, Add-AppxPackageVolume cmdlet, etc). In Settings when you change where new apps are saved to that's altering the default package volume for the machine. Under the covers Settings is using PackageManager's PackageVolume APIs to create a directory if you pick something other than C:. NOTE: Settings creates package volumes with a specific path but that's just Settings, the PackageVolume APIs let you specify a path (`Q:\Foo` is valid) and there's no restriction to # of package volumes on a volume (you could create `D:\Eeeny`, `D:\Meeny`, `D:\Miny` and `D:\Moe` if you chose).

Visual Studio F5 uses PackageManager.RegisterPackage*(...[DeploymentOption_DevelopmentMode](https://docs.microsoft.com/uwp/api/windows.management.deployment.deploymentoptions?view=winrt-22000)...) to register a project's output directory as the package directory. Upside is the package files are already there from building the project so no need to copy files elsewhere (makes for faster F5 experience). It's also NOT locked down so you could write to it if you chose. Of course that'll only work when running under VS... NOTE: The DevelopmentMode option has caveats, as it's designed for development not general use. See the docs for more details.

Packages can be installed with an External Location via `PackageManager.AddPackageByUriAsync(...new AddPackageOptions(){ [ExternalLocation](https://docs.microsoft.com/uwp/api/windows.management.deployment.addpackageoptions.externallocationuri?view=winrt-22000)="F:\Foo"...})`. In this case the package's metadata files (`appxmanifest.xml`, `appxblockmap.xml`, `signature.p7x` aka 'footprint' files) are in the specified/default package volume but the _content_ files (your EXEs, DLLs, images, data files, resources.pri, etc everything else) are in the External Location. That directory's outside deployment's control; it's referenced but _you_ are responsible for its content and lifetime. This is unique in that what's thought of as the "pkgdir" is split separating 'footprint' files (under packagevolume) vs 'content' files (per specified path). See [Grant identity to non-packaged desktop app](https://docs.microsoft.com/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps) for more info.

NOTE: PackageVolumes aren't writeable but that's necessary to bring benefits. Aside from higher integrity due to their no-write access, deployment's minimizing disk usage via shared-disk-space-for-identical-file-content is made possible by PackageVolumes. That's the only location providing the guarantees to safely share disk space by duplicate files.

"Inbox"/System packages that are built by and are a part of Windows (e.g. Settings) are owned by Windows and locked down (e.g. like ntdll.dll) but are in their own locations (usually somewhere under `C:\Windows`). These aren't under any PackageVolume.

Packages installed with [StageInPlace](https://docs.microsoft.com/uwp/api/windows.management.deployment.addpackageoptions.stageinplace?view=winrt-22000#windows-management-deployment-addpackageoptions-stageinplace) are another variation. Here you've specified a path where the pkgdir lives. So like the External Location _you_ are responsible for its lifetime and access, but unlike ExternalLocation there's just the 1 path containing all (content+footprint) files. StageInPlace is similar in effect to Inbox/system packages in that it's _a_ path merely _referenced_ by deployment but owned by someone else, though the mechanisms how system packages work differs from StageInPlace.

So to summarize:

* AddPackage -- pkgdir=package volume
* StageInPlace -- pkgdir=path you specify/control/manage/own
* ExternalLocation -- pkgdir=packagevolume for 'footprint' files, path you specify/control/manage/own for 'content' files
* Inbox/System -- pkgdir=path Windows specifes/controls/manages/owns
* RegisterPath(...DevelopmentMode...) -- pkgdir=path you specify/control/manage/own

**Comment by DrusTheAxe** [comment]
Packaged apps are installed to a location that's very locked down (e.g. `C:\Program Files\WindowsApps\Foo_1.2.3.4_x86__1234567890abc)` with ACLs only granting read (and maybe execute) rights but not write. This both protects the bits and is relied on by the sharing-across-users de-dupe'ing (when a package is registered to 5 people there's only 1 copy on disk, not 5). But that does mean a process can't e.g. `File.Open(GetPath(current-exe) + "\stuff").Write(stuff)`

`ApplicationData` is the right place for a packaged app to write data.

**Reply by Slion** [reply]
> Packaged apps are installed to a location that's very locked down 

True, even after admin elevation I could not modify a config file in there. At least on that machine that was causing problem.

**Reply by DrusTheAxe** [reply]
> > Packaged apps are installed to a location that's very locked down
> 
> True, even after admin elevation I could not modify a config file in there

Hence the _very_ in very locked down :-)

See above for more details.
